File: sdk/go/frame.go
package rtvbp

import (
	"encoding/json"
	"time"
)

// Kind identifies the semantic control-frame shape above an envelope codec.
type Kind uint8

const (
	KindRequest Kind = iota + 1
	KindResponse
	KindEvent
)

// ControlFrame is the envelope-independent semantic unit consumed by the session runtime.
// Only an Envelope codec converts this value to or from wire bytes.
type ControlFrame struct {
	// Kind identifies whether this is a request, response, or event.
	Kind Kind
	// ID identifies a request or event. It is empty when the envelope has no message identifier.
	ID string
	// CorrelID identifies the request answered by a response.
	CorrelID string
	// Method is the request method or event name.
	Method string
	// Payload contains request params, a response result, or event data.
	Payload json.RawMessage
	// Err is set only for an error response.
	Err *WireError
	// ReceivedAt is assigned by the transport after receipt, not by the envelope codec.
	ReceivedAt time.Time
}

// WireError is an envelope-independent response error.
type WireError struct {
	Code    int
	Message string
	Data    json.RawMessage
}

// Envelope encodes and decodes semantic control frames. Implementations are stateless and pure.
type Envelope interface {
	Name() string
	Encode(ControlFrame) ([]byte, error)
	Decode([]byte) (ControlFrame, error)
}

File: sdk/rust/src/frame.rs
use std::time::SystemTime;

use serde_json::Value;

/// The semantic control-frame kind above an envelope codec.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum FrameKind {
    Request,
    Response,
    Event,
}

/// One envelope-independent control frame.
#[derive(Clone, Debug, PartialEq)]
pub struct ControlFrame {
    pub kind: FrameKind,
    pub id: String,
    pub correlation_id: String,
    pub method: String,
    pub payload: Option<Value>,
    pub error: Option<WireError>,
    pub received_at: Option<SystemTime>,
}

impl ControlFrame {
    pub fn request(
        id: impl Into<String>,
        method: impl Into<String>,
        payload: Option<Value>,
    ) -> Self {
        Self {
            kind: FrameKind::Request,
            id: id.into(),
            correlation_id: String::new(),
            method: method.into(),
            payload,
            error: None,
            received_at: None,
        }
    }

    pub fn response(
        correlation_id: impl Into<String>,
        payload: Option<Value>,
        error: Option<WireError>,
    ) -> Self {
        Self {
            kind: FrameKind::Response,
            id: String::new(),
            correlation_id: correlation_id.into(),
            method: String::new(),
            payload,
            error,
            received_at: None,
        }
    }

    pub fn event(id: impl Into<String>, event: impl Into<String>, payload: Option<Value>) -> Self {
        Self {
            kind: FrameKind::Event,
            id: id.into(),
            correlation_id: String::new(),
            method: event.into(),
            payload,
            error: None,
            received_at: None,
        }
    }
}

/// An envelope-independent response error.
#[derive(Clone, Debug, PartialEq)]
pub struct WireError {
    pub code: i64,
    pub message: String,
    pub data: Option<Value>,
}

impl std::fmt::Display for WireError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}: {}", self.code, self.message)
    }
}

/// A stateless generated envelope codec.
pub trait Envelope: Send + Sync {
    fn name(&self) -> &'static str;

    /// Encode one semantic control frame.
    ///
    /// # Errors
    ///
    /// Returns an envelope error when the frame cannot be represented.
    fn encode(&self, frame: &ControlFrame) -> Result<Vec<u8>, crate::Error>;

    /// Decode one complete envelope message.
    ///
    /// # Errors
    ///
    /// Returns an envelope error when the bytes are malformed or incomplete.
    fn decode(&self, bytes: &[u8]) -> Result<ControlFrame, crate::Error>;
}
